src/perl.c

```diff
@@ -2117,12 +2159,30 @@ static int perl_shutdown (void)
 	t = perl_threads->tail;
 
 	while (NULL != t) {
+		struct timespec ts_wait;
 		c_ithread_t *thr = t;
 
 		/* the pointer has to be advanced before destroying
 		 * the thread as this will free the memory */
 		t = t->prev;
 
+		thr->shutdown = 1;
+		if (thr->running) {
+			/* Give some time to thread to exit from pi */
+			WARNING ("perl shutdown: thread is running inside perl. Waiting.");
+			ts_wait.tv_sec = 0;
+			ts_wait.tv_nsec = 500000;
+			nanosleep (&ts_wait, NULL);
```

Comment:
I don't feel like this is the right approach. Instead of sleeping an arbitrary time here and then blindly killing stuff below, I tend to think that we should just wait for the plugin to finish. I believe we do the same for "regular" read plugins. Alternatively, did you check if there's a way to cleanly shut down running perl code?